### PR TITLE
[Collective] Supporting multistream, stream pool, and CUDA events.

### PR DESCRIPTION
**Pull request by zhisbug**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/collective/collective_group/nccl_collective_group.py [hunk 1]
```diff
@@ -122,6 +160,10 @@ def __init__(self, world_size, rank, group_name):
         # record the used GPU IDs.
         self._used_gpu_indices = set()
 
+        # TODO(Fu): might need an event map
+        self._dev_pool_map = {}
```
Comment (zhisbug) [inline python/ray/util/collective/collective_group/nccl_collective_group.py:164]: this `_dev_pool_map` is never being managed...

File: python/ray/util/collective/collective_group/nccl_collective_group.py [hunk 2]
```diff
@@ -406,25 +449,44 @@ def _get_nccl_collective_communicator(self, comm_key, device_list):
...
         nccl_util.groupStart()
         for i, device in enumerate(device_list):
             actual_rank = self.rank * len(device_list) + i
             with nccl_util.Device(device):
                 comms[i] = nccl_util.create_nccl_communicator(
                     actual_world_size, nccl_uid, actual_rank)
-                streams[i] = cupy.cuda.Stream.null
-                # Stream(non_blocking=True)
+                if _MULTI_STREAM:
+                    pool: StreamPool = self._dev_pool_map.get(comm_key, StreamPool(self.POOL_SIZE))
```
Comment (zhisbug) [inline python/ray/util/collective/collective_group/nccl_collective_group.py:460]: every time when you try to query `_dev_pool_map` with `comm_key`, it is nonexistent, so you create a new Stream Pool? This is incorrect...

File: python/ray/util/collective/collective_group/nccl_collective_group.py [hunk 3]
```diff
@@ -406,25 +449,44 @@ def _get_nccl_collective_communicator(self, comm_key, device_list):
...
         # cupy allocate tensors on null streams.
-        cupy.cuda.Stream.null.synchronize()
+        # TODO(Fu): nccl document says we need recordStream besides calling this function
+        
+        if _MULTI_STREAM: 
+            for i, device in enumerate(device_list):
+                with nccl_util.Device(device):
+                    stream: cupy.cuda.Stream = streams[i]
+                    event: cupy.cuda.Event = events[i]
+                    event.record(cupy.cuda.get_current_stream())
```
Comment (zhisbug) [inline python/ray/util/collective/collective_group/nccl_collective_group.py:485]: what does `event.record` do here?

File: python/ray/util/collective/collective_group/nccl_collective_group.py [hunk 4]
```diff
@@ -481,10 +543,19 @@ def _get_nccl_p2p_communicator(self, comm_key, my_gpu_idx, peer_rank,
         # create the p2p communicators
         with nccl_util.Device(my_gpu_idx):
             comm = nccl_util.create_nccl_communicator(2, nccl_uid, my_p2p_rank)
-            stream = cupy.cuda.Stream.null
-            # Stream(non_blocking=True)
+            #TODO(Fu): get the stream from pool
+            if _MULTI_STREAM:
+                pool: StreamPool = self._dev_pool_map.get(comm_key, StreamPool(self.POOL_SIZE))
```
Comment (zhisbug) [inline python/ray/util/collective/collective_group/nccl_collective_group.py:548]: same here. You don't have code to manage the stream pool for each device at all. The current code will create stream pools every time when you try to query. This pool makes no sense.

File: python/ray/util/collective/collective_group/nccl_collective_group.py [hunk 5]
```diff
@@ -109,7 +111,43 @@ def get_nccl_id(self, timeout_s=180):
...
+            stream (cupy.cuda.Stream): the stream on which the kernel should be placed
+        """
+
+        self.mux.acquire()
+        # Create a new stream if it hasn't been initialized yet
+        if self.pool[self.index] == None:
+            # TODO(Fu): double check whether to set the non_blocking variable
+            self.pool[self.index] = cupy.cuda.Stream(null=False)
+        stream = self.pool[self.index]
+        self.index = (self.index + 1) % self.size
```
Comment (zhisbug) [inline python/ray/util/collective/collective_group/nccl_collective_group.py:149]: this way of deciding stream assignment seems too simplified. I believe in PyTorch there is a more complex mechanism for stream prioritization?

**Comment by zhisbug** [comment]
@richardliaw good to merge now.

**Review by richardliaw — APPROVED** [review]
1 nit

File: python/ray/util/collective/collective_group/cuda_stream.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                    logger.debug("NCCL multistream disabled.")
+                    self._pool[i] = cupy.cuda.Stream.null
+        self._init_flag = True
+
+
+# This is a map from GPU index to its stream pool.
+# It is supposed to be READ-ONLY out of this file
+_device_stream_pool_map = dict()
+for i in range(MAX_GPU_PER_ACTOR):
+    _device_stream_pool_map[i] = StreamPool(i)
```
[the remark this answers is not in the document]
Comment (zhisbug) [reply python/ray/util/collective/collective_group/cuda_stream.py]: addressed

**Pull request merged**